cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/AppFabricServer.java

```diff
@@ -181,6 +198,70 @@ protected void startUp() throws Exception {
                                                                       applicationCount);
     metricsCollectionService.getContext(Collections.emptyMap()).gauge(Constants.Metrics.Program.NAMESPACE_COUNT,
                                                                       namespaceCount);
+    // Fetch all apps to emit metadata for plugins
+    // This is used to back-fill plugin metadata after update
+    Map<ApplicationId, ApplicationMeta> apps =
+      TransactionRunners
+        .run(transactionRunner, (TxCallable<Map<ApplicationId, ApplicationMeta>>) context -> AppMetadataStore
+          .create(context).getAllApplications());
+
+    getPluginCounts(apps);
+  }
+
+  private void getPluginCounts(Map<ApplicationId, ApplicationMeta> apps)
+    throws IOException {
+    List<MetadataMutation> updates = new ArrayList<>();
+    for (Map.Entry<ApplicationId, ApplicationMeta> app : apps.entrySet()) {
+      collectPluginMetadata(app.getKey(), app.getValue().getSpec(), updates);
+    }
+    metadataStorage.batch(updates, MutationOptions.DEFAULT);
+  }
+
+  /**
+   * Helper method to emit plugin metdata for a given application
+   *
+   * @param applicationId ID of the application
+   * @param appSpec Application specification for the app
+   * @param updates List to keep track of Metadata Mutations that need to be executed
+   */
+  private void collectPluginMetadata(ApplicationId applicationId, ApplicationSpecification appSpec,
+                                     List<MetadataMutation> updates) {
+    String namespace = applicationId.getNamespace();
+    Map<MetadataEntity, Long> pluginCounts = appSpec.getPlugins().values().stream()
+      .map(p -> pluginToMetadataEntity(namespace, p))
+      .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+    String appKey = String.format("%s:%s", namespace, appSpec.getName());
+    for (Map.Entry<MetadataEntity, Long> entry : pluginCounts.entrySet()) {
+      LOG.trace("Adding application {} to plugin metadata for {}", appKey,
+                entry.getKey().getValue(MetadataEntity.PLUGIN));
+      updates.add(new MetadataMutation.Update(entry.getKey(),
```

Comment:
Seems like this is an update per plugin per application? This could potentially be a pretty large number, may be better to submit a batch of updates per app instead of collecting them all into a single list.

There's also race conditions to consider here. For example, suppose the read all the apps, somebody deletes one of the apps, then we update the plugin metadata. What happens in that scenario?

Reply:
Noted regarding the large batch size.

On the race condition, this should only happen once during startup. My assumption was that the user couldnt interact with CDAP during this time so no apps can be created/deleted. Is that not accurate?

Reply:
it's not uncommon to have thousands of apps, we wouldn't want to make CDAP unusable while we scan all of them and write metadata. Ideally we wouldn't scan for each startup either, just the first time 6.4.0 runs.